src/core/ext/filters/client_channel/retry_filter.cc

```diff
@@ -1713,6 +1689,34 @@ void RetryFilter::CallData::SetPollent(grpc_call_element* elem,
   calld->pollent_ = pollent;
 }
 
+void RetryFilter::CallData::PreCancel(grpc_call_element* elem,
+                                      grpc_error_handle error) {
+  auto* calld = static_cast<CallData*>(elem->call_data);
+  {
+    MutexLock lock(&calld->timer_mu_);
+    if (calld->retry_timer_pending_) {
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_retry_trace)) {
+        gpr_log(GPR_INFO, "calld=%p: cancelling retry timer: %s", calld,
+                grpc_error_string(error));
+      }
+      GRPC_ERROR_UNREF(error);
+      calld->retry_timer_pending_ = false;  // Checked by OnRetryTimer().
+      grpc_timer_cancel(&calld->retry_timer_);
+      calld->FreeAllCachedSendOpData();
+      GRPC_CALL_COMBINER_STOP(calld->call_combiner_, "Canceller");
+      return;
+    }
+  }
+  // Retry timer not pending, so propagate pre-cancellation down to LB call.
+  MutexLock lock(&calld->lb_call_pre_cancellation_mu_);
+  calld->lb_call_pre_cancelled_ = true;
+  if (calld->committed_call_ != nullptr) {
```

Comment:
nit: should there be a debug assert checking that only one of `calld->committed_call_` and `calld->call_attempt_` can be not null?

Reply:
Done.